src/sentry/models/externalactor.py

```diff
@@ -0,0 +1,36 @@
+import logging
+
+from django.db import models
+
+from sentry.db.models import BoundedPositiveIntegerField, DefaultFieldsModel, FlexibleForeignKey
+from sentry.types.integrations import ExternalProviders
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalActor(DefaultFieldsModel):
+    __core__ = False
+
+    actor = FlexibleForeignKey("sentry.Actor", db_index=True, on_delete=models.CASCADE)
+    organization = FlexibleForeignKey("sentry.Organization")
+    integration = FlexibleForeignKey("sentry.Integration", null=True)
+    provider = BoundedPositiveIntegerField(
+        choices=(
+            (ExternalProviders.EMAIL, "email"),
+            (ExternalProviders.SLACK, "slack"),
+            (ExternalProviders.MSTEAMS, "msteams"),
+            (ExternalProviders.PAGERDUTY, "pagerduty"),
+            (ExternalProviders.GITHUB, "github"),
+            (ExternalProviders.GITLAB, "gitlab"),
+        ),
+    )
+
+    # external name => display name i.e. username, team name, channel name
+    # external id => unique identifier i.e user id, channel id
+    external_name = models.TextField()
+    external_id = models.TextField(null=True)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_externalactor"
+        unique_together = (("organization", "provider", "external_name", "actor"),)
```

Comment:
Is there ever going to be a case where we have 2 slack integrations, that user has the same username in both workspaces and we want to have the user linked to both of these?

Reply:
@wedamija I think it's possible to have two slack integrations where the user has the same username for both workspaces, but I was under the impression that we didn't necessarily need to support that right now. 

From my understanding, we would have to do the following since the integration is nullable: 

```python
unique_together = (("organization", "provider", "external_name", "actor"), ("organization", "provider", "external_name", "actor", "integration"))
```

and that seemed fine to do later if we needed and we thought there was actually enough of use case for this. Looking at our data there are ~500 organizations that have more than 1 slack integration

@mgaeta @ceorourke  does that seem accurate to you both?

Reply:
That's `unique_together` won't actually work - `("organization", "provider", "external_name", "actor")` will take precedence and prevent and duplicate rows on that key, so the less restrictive `unique_together` won't do anything.

You need partials, something like `CREATE UNIQUE INDEX ON <table> (organization_id, provider, external_name, actor_id) where integration_id is null` and `CREATE UNIQUE INDEX ON <table> (organization_id, provider, external_name, actor_id, integration_id) where integration_id is not null`.

Alternatively, if there was some way to encapsulate `provider, integration_id` in another table that would help, since you'd be removing the nulls from this table. I don't know if something along that lines makes sense though, just throwing out a general thought.

Reply:
so even though I was wrong about that `unique_together`, I think this still stands:

> I think it's possible to have two slack integrations where the user has the same username for both workspaces, but I was under the impression that we didn't necessarily need to support that right now.

We can deal with adding a constraint later when/if needed, but right now are okay with this limitation. Hopefully we can zoom boom boom to django 2.2 and use https://docs.djangoproject.com/en/2.2/ref/models/constraints/#django.db.models.UniqueConstraint.condition